src/client/websocket/WebSocketShard.js

```diff
@@ -417,11 +406,49 @@ class WebSocketShard extends EventEmitter {
         this.ackHeartbeat();
         break;
       case OPCodes.HEARTBEAT:
-        this.sendHeartbeat();
+        this.sendHeartbeat('HeartbeatRequest');
         break;
       default:
         this.manager.handlePacket(packet, this);
+        if (this.status === Status.WAITING_FOR_GUILDS && packet.t === WSEvents.GUILD_CREATE) {
+          this.expectedGuilds.delete(packet.d.id);
+          this.checkReady();
+        }
+    }
+  }
+
+  /**
+   * Checks if the shard can be marked as ready
+   * @private
+   */
+  checkReady() {
+    // Step 0. Clear the ready timeout, if it exists
+    if (this.readyTimeout) this.manager.client.clearTimeout(this.readyTimeout);
+    // Step 1. If we don't have any other guilds pending, we are ready
+    if (!this.expectedGuilds.size) {
+      this.debug('Shard received all its guilds. Marking as fully ready.');
+      this.status = Status.READY;
+
+      /**
+       * Emitted when the shard is fully ready.
+       * This event is emitted if:
+       * * all guilds were received by this shard
+       * * the ready timeout expired, and some guilds are unavailable
+       * @event WebSocketShard#allReady
+       * @param {?Set<string>} unavailableGuilds Set of unavailable guilds, if any
+       */
+      this.emit(ShardEvents.ALL_READY);
+      return;
     }
+    // Step 2. Create a 15s timeout that will mark the shard as ready if there are still unavailable guilds
+    this.readyTimeout = this.manager.client.setTimeout(() => {
+      this.debug(`Shard did not receive any more guild packets in 15 seconds.
+  Unavailable guild count: ${this.expectedGuilds.size}`);
+
+      this.status = Status.READY;
+
+      this.emit(ShardEvents.ALL_READY, this.expectedGuilds);
+    }, 15000);
```

Comment:
`this.readyTimeout` is only used thrice in this file, here in line 444 and in https://github.com/discordjs/discord.js/blob/e47fc1e0f85ee63409979084bd45e0b748899fad/src/client/websocket/WebSocketShard.js#L426

Which checks if it's truthy (since it's either undefined or `NodeJS.Timeout`), but once this is set, it's never set to `undefined`, thus running `clearTimeout` every time `checkReady` is called and this has been set.

My guess is that you can assign `this.readyTimeout = undefined;` inside the callback so it is freed automatically, here as well as in line 426.